ethergo/submitter/submitter.go

```diff
@@ -346,119 +346,211 @@ func (t *txSubmitterImpl) txTypeForChain(chainID *big.Int) (txType uint8) {
 		}
+	}
+}
 
+// applyGasFloor applies the min gas price from the config if values are unset.
+// Otherwise, gas values are bumped by the configured GasBumpPercentage.
+//
+//nolint:cyclop,nestif
+func (t *txSubmitterImpl) applyGasFloor(ctx context.Context, transactor *bind.TransactOpts, chainID int, useDynamic, shouldBump bool) {
+	_, span := t.metrics.Tracer().Start(ctx, "submitter.applyGasFloor")
+
+	defer func() {
 		span.SetAttributes(
 			attribute.String("gas_price", bigPtrToString(transactor.GasPrice)),
 			attribute.String("gas_fee_cap", bigPtrToString(transactor.GasFeeCap)),
 			attribute.String("gas_tip_cap", bigPtrToString(transactor.GasTipCap)),
+			attribute.Bool("should_bump", shouldBump),
 		)
-		metrics.EndSpanWithErr(span, err)
+		metrics.EndSpan(span)
 	}()
 
-	// TODO: cache both of these values
-	shouldBump := true
-	useEIP1559 := t.config.SupportsEIP1559(chainID)
-	if useEIP1559 {
-		transactor.GasFeeCap, err = client.SuggestGasPrice(ctx)
-		if err != nil {
-			return fmt.Errorf("could not get gas price: %w", err)
+	gasFloor := t.config.GetMinGasPrice(chainID)
+	if shouldBump {
+		bumpPct := t.config.GetGasBumpPercentage(chainID)
+		if useDynamic {
+			if transactor.GasFeeCap == nil {
+				transactor.GasFeeCap = gasFloor
+				span.AddEvent("gas fee cap is nil; setting to gas floor")
+			} else {
+				transactor.GasFeeCap = gas.BumpByPercent(transactor.GasFeeCap, bumpPct)
+			}
+			if transactor.GasTipCap == nil {
+				transactor.GasTipCap = gasFloor
+				span.AddEvent("gas tip cap is nil; setting to gas floor")
+			} else {
+				transactor.GasTipCap = gas.BumpByPercent(transactor.GasTipCap, bumpPct)
+			}
+		} else {
+			transactor.GasPrice = gas.BumpByPercent(transactor.GasPrice, bumpPct)
 		}
-		// don't bump fee cap if we hit the max configured gas price
-		if transactor.GasFeeCap.Cmp(maxPrice) > 0 {
-			transactor.GasFeeCap = maxPrice
-			shouldBump = false
-			span.AddEvent("not bumping fee cap since max price is reached")
+	} else {
+		if useDynamic {
+			if transactor.GasFeeCap == nil || transactor.GasFeeCap.Cmp(gasFloor) < 0 {
+				transactor.GasFeeCap = gasFloor
+			}
+			if transactor.GasTipCap == nil || transactor.GasTipCap.Cmp(gasFloor) < 0 {
+				transactor.GasTipCap = gasFloor
+			}
+		} else if transactor.GasPrice == nil || transactor.GasPrice.Cmp(gasFloor) < 0 {
+			transactor.GasPrice = gasFloor
 		}
+	}
+}
+
+// applyGasFromOracle fetches gas values from a RPC endpoint and attempts to set them.
+// If values are already specified, they will be overridden if the oracle values are higher.
+func (t *txSubmitterImpl) applyGasFromOracle(ctx context.Context, transactor *bind.TransactOpts, client client.EVM, useDynamic bool) (err error) {
+	ctx, span := t.metrics.Tracer().Start(ctx, "submitter.applyGasFromOracle")
+
+	defer func() {
+		metrics.EndSpanWithErr(span, err)
+	}()
 
-		transactor.GasTipCap, err = client.SuggestGasTipCap(ctx)
+	if useDynamic {
+		suggestedGasFeeCap, err := client.SuggestGasPrice(ctx)
```

Comment:
I suggest we use some kind of multiplier here. `SuggestGasPrice` returns base fee from the latest block + suggestedGasTip iirc, so a slight increase in the base fee will lead to tx not being included (as usually the gas tip is a tiny fraction of GasFeeCap).

Reply:
Something like 1.5x-2x should work w/o forcing to spend extra, as you always get charged with _actual base fee_ + gas tip (that is unaffected by multiplier)

Reply:
Will circle back in this issue: https://github.com/synapsecns/sanguine/issues/2464